File: web_version/pmugg_files/shapes3D/shape3D.js
ms.shape3D = function(vertexTypes, edgeTypes, faceTypes, xml) {
	this.vertexTypes = vertexTypes;
	this.edgeTypes = edgeTypes;
	this.splicedEdgeTypes = {};
	this.faceTypes = faceTypes;
	this.is3D = true;
	this.xml = xml;
};

ms.shape3D.prototype.export = function() {
	var exp = ((x) => x.export(this));
	return {
		vertexTypes: this.vertexTypes.map(exp),
		edgeTypes: this.edgeTypes.map(exp),
		faceTypes: this.faceTypes.map(exp),
		xml: this.xml,
	};
};

ms.shape3D.import = function(json) {
	var result = new ms.shape3D([], [], [], json.xml);
	result.faceTypes = json.faceTypes.map((type) => ms.faceType3D.import(type));
	result.edgeTypes = json.edgeTypes.map((type) => ms.edgeType3D.import(type, result));
	result.vertexTypes = json.vertexTypes.map((type) => ms.vertexType.import(type, result));
	return result;
};

ms.shape3D.prototype.setNetworks = function(networks) {
	this.chainMap = networks.chainMap;
	this.vertex = networks.vertex;
	this.edge = networks.edge;
	this.face = networks.face;
};

ms.shape3D.prototype.getShape = function() {
	return this;
};

ms.shape3D.prototype.getVertices = function() {
	return [0];
};

ms.shape3D.prototype.getSpliceEdgeType = function(faceType, dir) {
	return ms.shape3D.getSpliceEdgeType(this.splicedEdgeTypes, this.edgeTypes, faceType, dir);
};

ms.shape3D.getSpliceEdgeType = function(splicedEdgeTypes, edgeTypes, faceType, dir) {
	var key = faceType.id + ',' + dir;
	if (!splicedEdgeTypes[key]) {
		var faceData = [{type: faceType, onRight: true}, {type: faceType, onRight: false}]
		var uv = (dir == 0) ? faceType.u : faceType.v;
		var edgeType = new ms.edgeType3D(faceData, uv);
		edgeType.setSpliced(true);
		splicedEdgeTypes[key] = edgeType;
		edgeTypes.push(edgeType);
	}
	return splicedEdgeTypes[key];
};


ms.shapeFile3D = function(name) {
	this.name = name;
	this.is3D = true;
};

ms.shapeFile3D.prototype.getName = function() {
	return this.name;
};

ms.shapeFile3D.prototype.isEmpty = function() {
	return false;
};

ms.shapeFile3D.prototype.parse = function() {
	var objFile = new OBJFile(ms.obj[this.name].obj, this.name);
	objFile.parse();
	this.model = objFile.result.models[0];


	// Switch from 1-index to 0-index.
	this.model.faces.forEach(function(face) {
		var n = face.vertices.length;
		for (var i = 0; i < n; i++) {
			// The indices start with 1.
			face.vertices[i].vertexIndex--;
			face.vertices[i].textureCoordsIndex--;
			face.vertices[i].vertexNormalIndex--;
		}
	});
};

File: cpp_version/pmugg/shapes3D/shape3d.cpp
#include "shape3d.h"

namespace ms {

Shape3D::Shape3D(const std::vector<VertexType*>& vTypes,
                 const std::vector<EdgeType3D*>& eTypes,
                 const std::vector<FaceType3D*>& fTypes,
                 const std::string& xmlData)
    : vertexTypes(vTypes)
    , edgeTypes(eTypes)
    , faceTypes(fTypes)
    , is3D(true)
    , xml(xmlData)
    // , chainMap(nullptr)
    // , vertex(nullptr)
    // , edge(nullptr)
    // , face(nullptr)
        {}

Shape3D::Shape3D() : Shape3D({}, {}, {}, "") {}

//std::map<std::string, std::vector<std::map<std::string, std::any>>> Shape3D::export_() const {
//    auto exporter = [this](const auto& x) { return x.export_(this); };
//    
//    return {
//        {"vertexTypes", std::vector<std::map<std::string, std::any>>(
//            vertexTypes.begin(), vertexTypes.end(), exporter)},
//        {"edgeTypes", std::vector<std::map<std::string, std::any>>(
//            edgeTypes.begin(), edgeTypes.end(), exporter)},
//        {"faceTypes", std::vector<std::map<std::string, std::any>>(
//            faceTypes.begin(), faceTypes.end(), exporter)},
//        {"xml", xml}
//    };
//}

Shape3D* Shape3D::import(const Json& json) {
    auto shape = new Shape3D();

    std::vector<FaceType3D*> faceTypes;
    for (const auto& type : json.at("faceTypes")) {
        faceTypes.push_back(FaceType3D::import(type));
    }
    shape->faceTypes = faceTypes;

    std::vector<EdgeType3D*> edgeTypes;
    for (const auto& type : json.at("edgeTypes")) {
        edgeTypes.push_back(EdgeType3D::import(type, shape));
    }
    shape->edgeTypes = edgeTypes;
    
    std::vector<VertexType*> vertexTypes;
    for (const auto& type : json.at("vertexTypes")) {
        vertexTypes.push_back(VertexType::import(type, shape));
    }
    shape->vertexTypes = vertexTypes;

    return shape;
}

Shape3D* Shape3D::getShape() {
    return this;
}

std::vector<int> Shape3D::getVertices() const {
    return {0};
}

//EdgeType3D* Shape3D::getSpliceEdgeType(
//    std::map<std::string, EdgeType3D*>& splicedEdgeTypes,
//    std::vector<EdgeType3D*>& edgeTypes,
//    FaceType3D* faceType,
//    int dir) {
//    
//    std::string key = std::to_string(faceType->id) + "," + std::to_string(dir);
//    
//    if (!splicedEdgeTypes.count(key)) {
//        std::vector<FaceData> faceData = {
//            {faceType, true},
//            {faceType, false}
//        };
//        
//        auto uv = (dir == 0) ? faceType->u : faceType->v;
//        auto edgeType = new EdgeType3D(faceData, uv);
//        edgeType->setSpliced(true);
//        
//        splicedEdgeTypes[key] = edgeType;
//        edgeTypes.push_back(edgeType);
//    }
//    
//    return splicedEdgeTypes[key];
//}
//
//EdgeType3D* Shape3D::getSpliceEdgeType(FaceType3D* faceType, int dir) {
//    return getSpliceEdgeType(splicedEdgeTypes, edgeTypes, faceType, dir);
//}

} // namespace ms 